tools/utils/download_support_methods.py

```diff
@@ -0,0 +1,117 @@
+    -------
+    datetime
+        searched month from where the availability of data starts for the given stock symbol and interval.
+    """
+    valid_date = end_date
+    while start_date <= end_date:
+        try:
+            total_days = (end_date - start_date).days
+            middle_date = start_date + timedelta(days=total_days // 2)
+            first_day = middle_date.replace(day=3)
+            last_day = (middle_date + timedelta(days=31)).replace(day=1) - timedelta(
+                days=1
+            )
+            stocks_url = (
+                f"http://127.0.0.1:8000/smart-api/equity/history/{stock_symbol}?interval={interval.name}"
+                f"&start_date={first_day.strftime('%Y-%m-%d')}%2009%3A15&end_date={last_day.strftime('%Y-%m-%d')}%2015%3A29"
+            )
+            response = requests.get(stocks_url, timeout=(60, 60))
+            if response.status_code == 200 and response.json():
+                end_date = first_day - timedelta(days=3)
+                valid_date = first_day.replace(day=1)
+            else:
+                start_date = last_day + timedelta(days=1)
+            time.sleep(0.2)
+        except Exception as e:
+            print(e)
+            start_date = last_day + timedelta(days=1)
+            continue
+    return valid_date
+
+
+def dataframe_to_json_files(
+    df: pd.DataFrame, dir_path: str, interval: CandlestickInterval
+):
+    """Process the given dataframe and convert it into suitable data structure i.e dictionary which will be stored in json file.
+
+    Parameters:
+    -----------
+    df: `pd.DataFrame`
+        pandas DataFrame to store into json files.
+    dir_path: `str`
+        Path of the destination directory to store json files.
+    """
+    # Convert timestamp to datetime and extract year and day
+    df["timestamp"] = pd.to_datetime(df["timestamp"])
+    df["year"] = df["timestamp"].dt.year
+    df["day"] = df["timestamp"].dt.strftime("%Y-%m-%d")
+    if interval.name == "ONE_DAY":
+        grouped = df.groupby("year")
+
+        # Iterate over each group
+        for year, group in grouped:
+            # Prepare data for JSON
+            data = (
+                group.set_index("day")
+                .drop(columns=["year", "timestamp"])
+                .to_dict(orient="index")
+            )
+            # Load
+            json_file_path = f"{dir_path}/{str(year)}.json"
+            if os.path.exists(json_file_path):
+                stored_data = load_json_data(json_file_path)
+                stored_data.update(data)
+                data = stored_data
+            write_to_json_file(json_file_path, data)
+    else:
+        df["time"] = df["timestamp"].dt.strftime("%H:%M")
+
+        # Group by year and day
+        grouped = df.groupby(["year", "day"])
+
+        # Iterate over each group
+        for (year, day), group in grouped:
+            # Create directory for the year if it doesn't exist
+            year_dir = f"{dir_path}/{str(year)}"
```

Comment:
Use path object instead of manually creating path like `year_dir = f"{dir_path}/{str(year)}"`. Because in windows the path connector is `\` instead of `/`, which will give error if we use it in windows.